### PR TITLE
📊 #2259 Core Completeness updates

### PR DESCRIPTION
**Pull request by demariadaniel**
**Description of changes**

Clinical Service changes related to Submitted Data Core Completeness UI
* Updates Core Completeness to include Normal Specimen / Tumour Specimen Data
* Moves Core Completeness into Donor entity data
* Fixes problems in dev/qa: Missing Treatment entity data, Filters empty entries, Fixes specimen/sample registration data

Related API PR https://github.com/icgc-argo/platform-api/pull/565
Related UI PR https://github.com/icgc-argo/platform-ui/pull/2329

**Type of Change**

- [ ] Bug
- [ ] Refactor
- [x] New Feature
- [ ] Release Candidate

**Checklist before requesting review:**

- [x] Check branch (code change PRs go to `develop` not master)
- [ ] Check copyrights for new files
- [x] Manual testing
- [x] Regression tests completed and passing (double check number of tests).
- [x] Spelling has been checked.
- [ ] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/clinical/service-worker-thread/tasks.ts [hunk 1]
```diff
@@ -71,27 +78,67 @@ const sortDocs = (sort: string) => (currentRecord: ClinicalInfo, nextRecord: Cli
...
+const mapEntityDocuments = (
+  totalDonors: number,
+  schemas: any,
+  query: ClinicalQuery,
+  completionStats: CompletionRecord[],
+) => (
+  entity: [string, ClinicalInfo[]],
+  index: number,
+  originalResultsArray: [string, ClinicalInfo[]][],
+) => {
```
Comment (joneubank) [inline src/clinical/service-worker-thread/tasks.ts:90]: Since this method is a bit complicated it would benefit from some comments explaining the work that is done in it and what the objective is. The return type of the inner method would be nice to declare as well.

File: src/clinical/service-worker-thread/tasks.ts [hunk 2]
```diff
@@ -71,27 +78,67 @@ const sortDocs = (sort: string) => (currentRecord: ClinicalInfo, nextRecord: Cli
...
-  results,
-]: [string, ClinicalInfo[]]) => {
-  if (isEmpty(results)) return undefined;
+const mapEntityDocuments = (
+  totalDonors: number,
+  schemas: any,
+  query: ClinicalQuery,
+  completionStats: CompletionRecord[],
+) => (
+  entity: [string, ClinicalInfo[]],
```
Comment (joneubank) [inline src/clinical/service-worker-thread/tasks.ts:87]: Why do entity as a tuple instead of a declared type like:
```
type EntityClinicalInfo = {
  entityName: string;
  clinicalInfo: ClinicalInfo[];
}
```

No performance difference but it will be a lot easier to keep track of the meaning of the two variables when they're named.

File: src/common-model/functions.ts [hunk 3]
```diff
@@ -133,27 +133,60 @@ export function getClinicalEntitySubmittedData(
...
+  // Sample Registration requires Specimen data
+  (entityTypes.includes('sampleRegistration') && !entityTypes.includes('specimens'))
+    ? ['completionStats', 'sampleRegistration', 'specimens']
+    : // Clinical Therapies require Treatments
+    // hormoneTherapy + treatment do not match schema names
+    entityTypes.includes('hormoneTherapy') ||
+      entityTypes.includes('treatment') ||
+      ClinicalTherapySchemaNames.some(entity => entityTypes.includes(entity))
+    ? ['treatments', 'treatment']
+    : '';
```
Comment (joneubank) [inline src/common-model/functions.ts:188]: Feels like the nested ternaries are getting a bit hard to parse... you might consider moving all the comments into a single block, and this might actually be a good use for an if/else structure since it would separate the conditions.

**Review by joneubank — APPROVED** [review]
Looks good, comments are just thoughts on some code style/readability that are my personal preferences.

File: src/clinical/service-worker-thread/tasks.ts (hunk 2 above)
Comment (joneubank) [reply src/clinical/service-worker-thread/tasks.ts]: Oh this is from the `Object.entries`, that explains it. I think the typing you've done here is fine, this makes more sense in context. If you add some comments on the method I think thats sufficient.

File: src/clinical/service-worker-thread/tasks.ts (hunk 1 above)
Comment (demariadaniel) [reply src/clinical/service-worker-thread/tasks.ts]: I ended up refactoring this a fair bit with comments and it should be much more readable

File: src/clinical/service-worker-thread/tasks.ts (hunk 2 above)
Comment (demariadaniel) [reply src/clinical/service-worker-thread/tasks.ts]: I ended up using an Object type + removing Object.entries which I'd been thinking of doing for a while. Also cleaned up the map function with new types added.

File: src/common-model/entities.ts [hunk 4]
```diff
@@ -39,7 +39,7 @@ export enum ClinicalEntitySchemaNames {
   BIOMARKER = 'biomarker',
 }
 
-export const aliasEntityNames = {
+export const aliasEntityNames: Record<string, string> = {
```
Comment (blabadi) [inline src/common-model/entities.ts:42]: it's better to utilize typescript type system than using strings
for example instead of `Record<string, string>` use `Record<EntityName, EntityAlias>` where each one of these is either an enum or literal type `type EntityAlias = 'donor' | 'sampleRegistration' | ....` or  use `keyof ...` 
the point is to avoid Strings in places where arbitrary strings can cause runtime errors

File: src/common-model/functions.ts [hunk 5]
```diff
@@ -133,27 +133,65 @@ export function getClinicalEntitySubmittedData(
...
             ...entity.clinicalInfo,
-          },
-    )
-    .filter(notEmpty);
+          }));
 
   return clinicalRecords;
 }
 
+export const requiredEntities = (entityTypes: string[]) => {
```
Comment (blabadi) [inline src/common-model/functions.ts:176]: same as above comment, also applies to return type

File: src/common-model/entities.ts (hunk 4 above)
Comment (demariadaniel) [reply src/common-model/entities.ts]: Nice I've made these updates, definitely better

File: src/common-model/functions.ts [hunk 6]
```diff
@@ -133,27 +133,65 @@ export function getClinicalEntitySubmittedData(
...
+    return ['completionStats', 'sampleRegistration', 'specimens'];
+  } else if (
+    // Clinical Therapies require Treatments
+    // hormoneTherapy + treatment do not match schema names
+    entityTypes.includes('hormoneTherapy') ||
+    entityTypes.includes('treatment') ||
+    ClinicalTherapySchemaNames.some(entity => entityTypes.includes(entity))
+  ) {
+    return ['treatments', 'treatment'];
+  } else return '';
```
Comment (blabadi) [inline src/common-model/functions.ts:193]: it's better to keep the return type consistent instead of empty string why not return an empty array

File: src/clinical/donor-repo.ts [hunk 7]
```diff
@@ -190,7 +183,11 @@ export const donorDao: DonorRepository = {
   ): Promise<DeepReadonly<{ donors: Donor[]; totalDonors: number }>> {
     const { sort, pageSize, entityTypes, donorIds, submitterDonorIds, completionState } = query;
 
-    const projection = [...DONOR_ENTITY_CORE_FIELDS, ...entityTypes].join(' ');
+    const projection = [
+      ...DONOR_ENTITY_CORE_FIELDS,
+      ...entityTypes,
+      ...requiredEntities(entityTypes),
```
Comment (blabadi) [inline src/clinical/donor-repo.ts:189]: ok so I see this function is to get the required document fields for the projection, I'd call it `getRequiredDonorDocumentFieldsForEntityTypes()` a bit more verbose but clearer

File: src/common-model/functions.ts [hunk 8]
```diff
@@ -133,27 +134,67 @@ export function getClinicalEntitySubmittedData(
...
+    return ['completionStats', 'sampleRegistration', 'specimens'];
+  } else if (
+    // Clinical Therapies require Treatments
+    // hormoneTherapy + treatment do not match schema names
+    entityTypes.includes('hormoneTherapy') ||
+    entityTypes.includes('treatment') ||
+    ClinicalTherapySchemaNames.some(entity => entityTypes.includes(entity))
+  ) {
+    return ['treatments'];
+  } else return [''];
```
Comment (blabadi) [inline src/common-model/functions.ts:195]: why not just return `[ ]` the empty string is not really needed, right ?

Comment (demariadaniel) [reply src/common-model/functions.ts:195]: Yes updating that next commit

File: src/common-model/functions.ts [hunk 9]
```diff
@@ -133,27 +134,67 @@ export function getClinicalEntitySubmittedData(
...
-    )
-    .filter(notEmpty);
+          }));
 
   return clinicalRecords;
 }
 
+export const requiredEntities = (
+  entityTypes: EntityAlias[],
+): EntityAlias | ClinicalEntitySchemaNames | string[] => {
```
Comment (blabadi) [inline src/common-model/functions.ts:179]: the return type can be simpler:
`Array<EntityAlias | 'completionStats'>`

File: src/clinical/donor-repo.ts (hunk 7 above)
Comment (demariadaniel) [reply src/clinical/donor-repo.ts:189]: Sure I'll use `getRequiredDonorFieldsForEntityTypes` just to keep it a little concise, I think 'document' is understood

File: src/common-model/functions.ts (hunk 6 above)
Comment (demariadaniel) [reply src/common-model/functions.ts]: Updated

File: src/clinical/service-worker-thread/tasks.ts [hunk 10]
```diff
@@ -19,22 +19,47 @@
...
 type RecordsMap = {
   [key in ClinicalEntitySchemaNames]: ClinicalInfo[];
 };
 
+type EntityClinicalInfo = {
+  entityName: ClinicalEntitySchemaNames;
+  results: ClinicalInfo[];
+};
+
+const queryEntityNames = <string[]>Object.values(aliasEntityNames);
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply src/clinical/service-worker-thread/tasks.ts]: Nice, updated

File: src/common-model/functions.ts (hunk 9 above)
Comment (demariadaniel) [reply src/common-model/functions.ts]: Updated to `Array<EntityAlias | ClinicalEntitySchemaNames | 'completionStats'>`
